### DISCUSSION TITLE
Web app questions

### DISCUSSION THREAD

**Discussion by kyleskom** [nealmick/bb#49] [Q&A]
1.) What did you use to build your front end?
2.) What do you use to host it

Just looking see how you built your web app, it looks great and am curious to learn.

**Comment by nealmick** [comment]
Hey, I'm using django for pretty much everything.  
It's a batteries included python web framework.  It routes urls and calls python code to render a page, its all the same as flask, just has more features.   I think your project has a single url route and it renders a single template.  Django does the same thing uses jinja to render html templates, soo basically it routes a url to some python called a view, which in tern passes some context to the template which is rendered and sent to browser.  Flask is better alternative for smaller projects, but i think if the project involves users then django is better and offers structure and more built in features.  Flask html templates are essentialy the same as django and both use the same templating language to write python in html.

for hosting im using Digital Ocean, just a marketplace image for django, but i did this back a long time ago and now i would definitely try and containerize the app with docker.  The main django server is a Digital Ocean droplet, it runs nginx to connect.  I had to move model training off the main server tho, it kept downing the server, and caused alot of load and lots of problems.  So I setup a separate cluster on fly.io just for training models, it has 2 django endpoints one for start training and one for live training logs.  Once training session completes it sends the model weights back to the main server where they are saved by the username.  The training cluster isn't anything crazy its a 4 core 4 gb ram, and it auto sleeps with quick starts so i don't have to pay constant rent on it and it will scale up if needed.  No gpu or anything, although fly does have nvidia h100 servers they really expensive and not needed.    The main server still handles most the stuff it does model inference and model retrains on single games, everything except the long training sessions. 

The front end is just html, vanilla css, and little bits and pieces of javascript to tie it together.  The home page is on a no code platform called webflow, its a template off the market place I think its called radiant.  Also using cloudflare to manage DNS.

Anyways hope this is helpful, let me know if you have any other question
